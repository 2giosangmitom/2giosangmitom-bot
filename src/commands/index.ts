export * as pingCmd from './ping';
export * as leetcodeCmd from './leetcode';
export * as waifuCmd from './waifu';
