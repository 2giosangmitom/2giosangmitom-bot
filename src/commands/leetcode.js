import { EmbedBuilder, MessageFlags, SlashCommandBuilder } from 'discord.js';
import { randomFrom } from '../lib/utils.js';
import { filterQuestions, loadData } from '../services/leetcode.js';
import Fuse from 'fuse.js';

const difficulties = ['Easy', 'Medium', 'Hard'];

// Cache the LeetCode data promise to avoid multiple API calls
/** @type {Promise<import('../services/leetcode.js').LeetCodeData | null> | null} */
let leetCodeDataPromise = null;
/** @type {Fuse<string> | null} */
let fuseInstance = null;

/**
 * Gets or initializes the LeetCode data with caching
 * @returns {Promise<import('../services/leetcode.js').LeetCodeData | null>}
 */
async function getLeetCodeData() {
  // Reset cache in test environment or if promise is stale
  if (!leetCodeDataPromise || process.env.NODE_ENV === 'test' || process.env.VITEST) {
    leetCodeDataPromise = loadData();
  }
  const result = await leetCodeDataPromise;
  // If data is null, reset cache for next call
  if (!result) {
    leetCodeDataPromise = null;
    fuseInstance = null; // Also reset Fuse instance
  }
  return result;
}

/**
 * Gets or initializes the Fuse.js instance for topic search
 * @returns {Promise<Fuse<string> | null>}
 */
const data = new SlashCommandBuilder()
  .setName('leetcode')
  .setDescription('Get random questions from LeetCode')
  .addStringOption((option) =>
    option
      .setName('difficulty')
      .setDescription('Set difficulty level')
      .addChoices(difficulties.map((v) => ({ name: v, value: v })))
  )
  .addStringOption((option) => option.setName('topic').setDescription('Set topic of the problem').setAutocomplete(true))
  .addBooleanOption((option) => option.setName('include-paid').setDescription('Include paid problems'));

/**
 * Executes the leetcode command
 * @param {import('discord.js').ChatInputCommandInteraction} interaction
 */
async function execute(interaction) {
  const leetcodeData = await getLeetCodeData();
  if (!leetcodeData) {
    throw new Error('No problems found at the moment');
  }

  let difficultyParam = interaction.options.getString('difficulty');
  let topicParam = interaction.options.getString('topic');
  const includePaidParam = interaction.options.getBoolean('include-paid') ?? false;

  // Set defaults if not provided
  if (!difficultyParam) {
    const randomDifficulty = randomFrom(difficulties);
    if (!randomDifficulty) {
      throw new Error('An error occurred while reading parameters');
    }
    difficultyParam = randomDifficulty;
  }

  if (!topicParam) {
    const randomTopic = randomFrom(leetcodeData.topics);
    if (!randomTopic) {
      throw new Error('An error occurred while reading parameters');
    }
    topicParam = randomTopic;
  }

  const filterdProblems = filterQuestions(
    leetcodeData.problems,
    difficultyParam.toLowerCase(),
    topicParam.toLowerCase(),
    includePaidParam
  );

  // Check if filtered problems array is empty
  if (filterdProblems.length === 0) {
    await interaction.reply({
      content: `No problems found matching the criteria: difficulty=${difficultyParam}, topic=${topicParam}, include-paid=${includePaidParam}`,
      flags: MessageFlags.Ephemeral
    });
    return;
  }

  const problem = randomFrom(filterdProblems);
  if (!problem) {
    throw new Error('No problems match your preference');
  }

  const embed = new EmbedBuilder()
    .setTitle(`${problem.title}`)
    .setURL(problem.url)
    .setColor(problem.difficulty === 'easy' ? 5763719 : problem.difficulty === 'medium' ? 16776960 : 15548997)
    .setFooter({ text: `Powered by LeetCode` })
    .setTimestamp()
    .addFields(
      {
        name: 'Acceptance Rate',
        value: `${(problem.acRate * 100).toFixed(2)}%`,
        inline: true
      },
      {
        name: 'Paid Only',
        value: problem.isPaid ? '✅ Yes' : '❌ No',
        inline: true
      },
      {
        name: 'Topics',
        value: problem.topics.length > 0 ? problem.topics.join(', ') : 'None'
      }
    );

  await interaction.reply({ embeds: [embed] });
}

/**
 * Handles autocomplete for the topic parameter
 * @param {import('discord.js').AutocompleteInteraction} interaction
 */
async function autocomplete(interaction) {
  const leetcodeData = await getLeetCodeData();
  if (!leetcodeData) {
    throw new Error('No problems found at the moment');
  }

  const focusedValue = interaction.options.getFocused();

  // If no search term, return first 25 topics
  if (!focusedValue || !focusedValue.trim()) {
    const topicsToShow = leetcodeData.topics.slice(0, 25);
    await interaction.respond(topicsToShow.map((topic) => ({ name: topic, value: topic })));
    return;
  }

  // Create Fuse instance for search if we have a search term
  if (!fuseInstance && leetcodeData.topics && leetcodeData.topics.length > 0) {
    fuseInstance = new Fuse(leetcodeData.topics, {
      includeScore: true,
      threshold: 0.3
    });
  }

  if (fuseInstance) {
    const result = fuseInstance.search(focusedValue, { limit: 25 });
    if (result && Array.isArray(result)) {
      await interaction.respond(result.map((res) => ({ name: res.item, value: res.item })));
    } else {
      await interaction.respond([]);
    }
  } else {
    await interaction.respond([]);
  }
}

export { data, execute, autocomplete };
