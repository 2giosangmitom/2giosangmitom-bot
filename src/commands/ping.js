import { SlashCommandBuilder } from 'discord.js';

const data = new SlashCommandBuilder().setName('ping').setDescription('Replies with Pong and latency info');

/**
 * @param {import('discord.js').ChatInputCommandInteraction} interaction
 */
async function execute(interaction) {
  await interaction.reply('Pinging...');

  const sent = await interaction.fetchReply();
  const roundTripLatency = sent.createdTimestamp - interaction.createdTimestamp;
  const wsPing = interaction.client.ws.ping;

  await interaction.editReply(
    `🏓 Pong!\n🏃 Round-trip latency: \`${roundTripLatency}ms\`\n🏃 WebSocket ping: \`${wsPing}ms\``
  );
}

export { data, execute };
