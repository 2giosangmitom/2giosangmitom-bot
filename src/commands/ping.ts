import { type ChatInputCommandInteraction, SlashCommandBuilder } from 'discord.js';

const data = new SlashCommandBuilder()
  .setName('ping')
  .setDescription('Get latency-related infomation');

/**
 * @description Get latency information and reply to user
 * @param interaction The slash command interaction object
 */
async function execute(interaction: ChatInputCommandInteraction) {
  await interaction.reply('Pinging...');

  const sent = await interaction.fetchReply();
  const roundTripLatency = sent.createdTimestamp - interaction.createdTimestamp;
  const wsPing = interaction.client.ws.ping;

  await interaction.editReply(
    `🏓 Pong!\n🏃 Round-trip latency: \`${roundTripLatency}ms\`\n🏃 WebSocket ping: \`${wsPing}ms\``
  );
}

export { data, execute };
