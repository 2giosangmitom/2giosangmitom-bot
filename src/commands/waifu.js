import { SlashCommandBuilder, EmbedBuilder } from 'discord.js';
import { getImage, categories } from '../services/waifu.js';
import { randomFrom } from '../lib/utils.js';

/** @type {string[]} */
const titles = [
  "Here's Your Daily Dose of Motivation ✨",
  'A Waifu Appears! 💖',
  'Stay Strong, Senpai! 💪',
  "You Got This! Here's Some Motivation 🔥",
  'Summoning Your Waifu... 💫',
  'Your Waifu Believes in You! 🌸',
  'Power Up Time! 🚀',
  "Don't Give Up, Senpai! 💥",
  'Waifu Buff Activated! ⚡',
  "Keep Going, You're Doing Great! 🌟",
  'One Step Closer to Victory! 🏆',
  'Level Up Your Spirit! 🆙',
  'Another Day, Another Quest! 🗺️',
  "You're Stronger Than You Think! 🐉",
  "Waifu's Blessing Incoming! 🍀"
];

const data = new SlashCommandBuilder()
  .setName('waifu')
  .setDescription('Get a random cute anime girl image to boost your motivation 💖')
  .addStringOption((option) =>
    option
      .setName('category')
      .setDescription('The category of image')
      .addChoices(categories.map((v) => ({ name: v, value: v })))
  );

/**
 * Executes the waifu command and returns a motivational anime image
 * @param {import('discord.js').ChatInputCommandInteraction} interaction - The command interaction
 * @returns {Promise<void>}
 */
async function execute(interaction) {
  await interaction.deferReply();

  try {
    const categoryParam = interaction.options.getString('category')?.toLowerCase() || undefined;
    const { url, category } = await getImage(categoryParam);

    const embed = new EmbedBuilder()
      .setColor(15431372)
      .setTitle(randomFrom(titles) || 'Your Waifu is Here! 💖')
      .setDescription(`*Category: ${category}*`)
      .setImage(url)
      .setFooter({ text: 'Powered by waifu.pics' })
      .setTimestamp();

    await interaction.editReply({ embeds: [embed] });
  } catch (error) {
    throw new Error(`Failed to fetch waifu image: ${error instanceof Error ? error.message : 'Unknown error'}`);
  }
}

export { data, execute };
