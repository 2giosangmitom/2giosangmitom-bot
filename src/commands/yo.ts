import { SlashCommandBuilder } from 'discord.js';
import type { SlashCommand } from '~/types';

const command: SlashCommand = {
  data: new SlashCommandBuilder()
    .setName('yo')
    .setDescription('Music commands')
    .addSubcommand((subcommand) =>
      subcommand
        .setName('play')
        .setDescription('Play a track from SoundCloud')
        .addStringOption((option) =>
          option.setName('query').setDescription('An URL or a search term').setRequired(true)
        )
    ),
  async execute(interaction) {
    const subcommand = interaction.options.getSubcommand();

    if (subcommand === 'play') {
      await interaction.deferReply();
      const queryParam = interaction.options.getString('query', true);
      await interaction.client.music.play(queryParam, interaction);
    }
  }
};

export default command;
