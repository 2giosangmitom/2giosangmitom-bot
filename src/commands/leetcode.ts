/**
 * @author Vo Quang Chien <voquangchien.dev@proton.me>
 * @license MIT
 * @copyright © 2025 Vo Quang Chien
 */

import {
  AutocompleteInteraction,
  ChatInputCommandInteraction,
  EmbedBuilder,
  SlashCommandBuilder
} from 'discord.js';
import Fuse from 'fuse.js';
import { randomFrom } from '~/lib/utils';
import { difficulties, loadData, filterQuestions } from '~/services/leetcode';

// LeetCode data cache
let leetcodeData: LeetCodeData | null = null;

// Fuse instance
let fuseInstance: Fuse<string> | null = null;

const data = new SlashCommandBuilder()
  .setName('leetcode')
  .setDescription('Get random questions from LeetCode')
  .addStringOption((option) =>
    option
      .setName('difficulty')
      .setDescription('Set difficulty level')
      .addChoices(difficulties.map((v) => ({ name: v, value: v })))
  )
  .addStringOption((option) =>
    option.setName('topic').setDescription('Set topic of the problem').setAutocomplete(true)
  )
  .addBooleanOption((option) =>
    option.setName('include-paid').setDescription('Include paid problems')
  );

/**
 * @description Get random image from waifu.pics and send to user.
 * @param interaction The slash command interaction object.
 */
async function execute(interaction: ChatInputCommandInteraction) {
  if (!leetcodeData) {
    if (!(leetcodeData = await loadData())) {
      throw new Error('No problems found at the moment');
    }
  }

  const difficultyParam = interaction.options.getString('difficulty') ?? undefined;
  const topicParam = interaction.options.getString('topic') ?? undefined;
  const includePaidParam = interaction.options.getBoolean('include-paid') ?? undefined;

  const questions = filterQuestions(leetcodeData, difficultyParam, topicParam, includePaidParam);
  const problem = randomFrom(questions);
  if (!problem) {
    throw new Error('No problems match your preference.');
  }

  const embed = new EmbedBuilder()
    .setTitle(`${problem.title}`)
    .setURL(problem.url)
    .setColor(
      problem.difficulty === 'easy' ? 'Green' : problem.difficulty === 'medium' ? 'Yellow' : 'Red'
    )
    .setFooter({ text: `Powered by LeetCode` })
    .setTimestamp()
    .addFields(
      {
        name: 'Acceptance Rate',
        value: `${(problem.acRate * 100).toFixed(2)}%`,
        inline: true
      },
      {
        name: 'Paid Only',
        value: problem.isPaid ? '✅ Yes' : '❌ No',
        inline: true
      },
      {
        name: 'Topics',
        value: problem.topics.length > 0 ? problem.topics.join(', ') : 'None'
      }
    );

  await interaction.reply({ embeds: [embed] });
}

/**
 * @description Handles autocomplete for the topic parameter
 */
async function autocomplete(interaction: AutocompleteInteraction) {
  if (!leetcodeData) {
    throw new Error('No problems found at the moment');
  }

  const focusedValue = interaction.options.getFocused();

  // If no search term, return first 25 topics
  if (!focusedValue || !focusedValue.trim()) {
    const topicsToShow = leetcodeData.topics.slice(0, 25);
    await interaction.respond(topicsToShow.map((topic) => ({ name: topic, value: topic })));
    return;
  }

  // Create Fuse instance for search if we have a search term
  if (!fuseInstance && leetcodeData.topics && leetcodeData.topics.length > 0) {
    fuseInstance = new Fuse(leetcodeData.topics, {
      includeScore: true,
      threshold: 0.3
    });
  }

  if (fuseInstance) {
    const result = fuseInstance.search(focusedValue, { limit: 25 });
    if (result && Array.isArray(result)) {
      await interaction.respond(result.map((res) => ({ name: res.item, value: res.item })));
    } else {
      await interaction.respond([]);
    }
  } else {
    await interaction.respond([]);
  }
}

export { data, execute, autocomplete };
