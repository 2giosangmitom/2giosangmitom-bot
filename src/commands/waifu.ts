/**
 * @author Vo Quang Chien <voquangchien.dev@proton.me>
 * @license MIT
 * @copyright Â© 2025 Vo Quang Chien
 */

import { ChatInputCommandInteraction, EmbedBuilder, italic, SlashCommandBuilder } from 'discord.js';
import { getImage, categories } from '~/services/waifu';

const data = new SlashCommandBuilder()
  .setName('waifu')
  .setDescription('Get a random cute anime girl image to boost your motivation ðŸ’–')
  .addStringOption((option) =>
    option
      .setName('category')
      .setDescription('The category of the image')
      .addChoices(categories.map((category) => ({ name: category, value: category })))
  );

/**
 * @description Get random image from waifu.pics and send to user.
 * @param interaction The slash command interaction object.
 */
async function execute(interaction: ChatInputCommandInteraction) {
  await interaction.deferReply();

  const categoryParam = interaction.options.getString('category') ?? undefined;

  const { url, category, title } = await getImage(categoryParam);

  const waifuEmbed = new EmbedBuilder()
    .setColor('LuminousVividPink')
    .setTitle(title)
    .setDescription(italic(`Category: ${category}`))
    .setImage(url)
    .setFooter({ text: 'Powered by waifu.pics' })
    .setTimestamp();

  await interaction.editReply({ embeds: [waifuEmbed] });
}

export { data, execute };
