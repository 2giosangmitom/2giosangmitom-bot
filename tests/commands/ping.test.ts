import { ChatInputCommandInteraction } from 'discord.js';
import { describe, it, expect, vi, beforeEach, Mock } from 'vitest';
import { execute } from '~/commands/ping';

interface MockInteraction {
  reply: Mock;
  fetchReply: Mock;
  client: {
    ws: {
      ping: number;
    };
  };
  editReply: Mock;
  createdTimestamp: number;
}

describe('ping', () => {
  let mockInteraction: MockInteraction;

  beforeEach(() => {
    mockInteraction = {
      reply: vi.fn(),
      fetchReply: vi.fn().mockReturnValue({ createdTimestamp: 1000 }),
      client: {
        ws: {
          ping: 100
        }
      },
      editReply: vi.fn(),
      createdTimestamp: 1000
    };
  });

  it('should reply with correct ping information', async () => {
    mockInteraction.fetchReply.mockReturnValueOnce({
      createdTimestamp: 1100 // 100ms later
    });

    await execute(mockInteraction as unknown as ChatInputCommandInteraction);

    expect(mockInteraction.reply).toBeCalledTimes(1);
    expect(mockInteraction.reply).toBeCalledWith('Pinging...');
    expect(mockInteraction.editReply).toBeCalledTimes(1);
    expect(mockInteraction.editReply).toBeCalledWith(
      '🏓 Pong!\n🏃 Round-trip latency: `100ms`\n🏃 WebSocket ping: `100ms`'
    );
  });

  it.each([
    { created: 1000, replied: 1050, expected: 50 },
    { created: 2000, replied: 2250, expected: 250 },
    { created: 5000, replied: 5001, expected: 1 }
  ])(
    'should reply $expected when created = $created and replied = $replied',
    async ({ created, replied, expected }) => {
      mockInteraction.createdTimestamp = created;
      mockInteraction.fetchReply.mockReturnValueOnce({ createdTimestamp: replied });

      await execute(mockInteraction as unknown as ChatInputCommandInteraction);

      expect(mockInteraction.editReply).toBeCalledWith(
        expect.stringContaining(`Round-trip latency: \`${expected}ms\``)
      );
    }
  );

  it.each([100, 200, 50, 20])(
    'should display WebSocket ping correctly when ping = %i',
    async (ping) => {
      mockInteraction.client.ws.ping = ping;

      await execute(mockInteraction as unknown as ChatInputCommandInteraction);

      expect(mockInteraction.editReply).toBeCalledWith(
        expect.stringContaining(`WebSocket ping: \`${ping}ms\``)
      );
    }
  );
});
