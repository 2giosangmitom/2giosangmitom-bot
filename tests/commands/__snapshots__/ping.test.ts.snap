// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ping > should reply with correct ping information 1`] = `
[
  [
    "🏓 Pong!
🏃 Round-trip latency: \`100ms\`
🏃 WebSocket ping: \`100ms\`",
  ],
]
`;
